Add guides for some integration scenarios (#337)
wso2/ballerina-integrator: docs/content/src/learn/guides/integration-patterns/scatter-gather-flow/README.md

# Scatter-Gather Flow Control

## About

Scatter-Gather is an integration pattern where a request is sent to multiple recipients and each of the responses are aggregated and returned back to the client as a single response. This guide demonstrates a simple scatter-gather scenario where two files in an FTP location are read simultaneously, their contents aggregated and sent back to the client.

## What you'll build

We shall have a service ‘employeeDetails’ which accepts a client request and reads two different CSV files with employee details from an FTP server. It then converts the responses of each read to json, aggregates both the json and sends back to the client.

## Prerequisites

- [Ballerina Distribution](https://ballerina.io/learn/getting-started/)
- A Text Editor or an IDE 
> **Tip**: For a better development experience, install the Ballerina IDE plugin for [VS Code](https://marketplace.visualstudio.com/items?itemName=ballerina.ballerina)
- An FTP Server

## Implementation

> * Create a new ballerina project named ‘scatter-gather-flow’.

Reviewer: ```suggestion
* Create a new Ballerina project named `scatter-gather-flow`.
```

Let's change "ballerina" to "Ballerina" everywhere. Also let's use backticks instead of single quotes for project names, file names, and anything related to the code.

* Now let's create a file called `ballerina.conf` under the root path of the project structure. The file should have the following configurations.

```
FTP_HOST = "<ftp_host>"
FTP_PORT = <port>
FTP_USER = "<username>"
FTP_PASSWORD = "<password>"
```

The username and password are sensitive data. Therefore those should not be hard-coded. Ballerina supports encrypting sensitive data and uses them in the program.

```shell
$ ballerina encrypt
```

The execution of the encrypt action will ask for value and secret key. Once you provide input, it'll output the encrypted value that can directly be used in the config file.

```shell
Enter value:

Enter secret:

Re-enter secret to verify:

Add the following to the configuration file:
<key>="@encrypted:{aoIlSvOPeBEZ0COma+Wz2uWznlNn1IWz4StiWQCO6g4=}"

Or provide it as a command line argument:
--<key>=@encrypted:{aoIlSvOPeBEZ0COma+Wz2uWznlNn1IWz4StiWQCO6g4=}
```

Now we can use the encrypted value in the `ballerina.conf` file.

```
FTP_USER = "@encrypted:{aoIlSvOPeBEZ0COma+Wz2uWznlNn1IWz4StiWQCO6g4=}"
FTP_PASSWORD = "@encrypted:{aoIlSvOPeBEZ0COma+Wz2uWznlNn1IWz4StiWQCO6g4=}"
```

> * Before writing the service, let's create two csv files `employees1.csv` and `employees2.csv` with the following content and upload to an FTP server.

Reviewer: ```suggestion
* Before writing the service, let's create two CSV files `employees1.csv` and `employees2.csv` with the following content and upload to an FTP server.
```

```csv
empId,firstName,lastName,joinedDate
100,Lois,Walker,11/24/2003
101,Brenda,Robinson,7/27/2008
102,Joe,Robinson,8/3/2016
103,Diane,Evans,4/16/1999
104,Benjamin,Russell,7/25/2013
```

```csv
empId,firstName,lastName,joinedDate
105,Nancy,Baker,7/22/2005
106,Carol,Murphy,9/14/2016
107,Frances,Young,1/28/1983
108,Diana,Peterson,4/27/1994
109,Ralph,Flores,2/17/2014
```

* Now open the main.bal file and add the following content. This is going to be our integration logic.

**main.bal**
<!-- INCLUDE_CODE: src/employee_details_service/main.bal -->

## Run the Integration

* First let’s build the module. While being in the scatter-gather-flow directory, execute the following command.

> ```bash

Reviewer: Can we improve formatting so that code block aligns with the bulleted point?

$ ballerina build employee_details_service
```

This would create the executables without running the test cases. 

> * Now run the jar file created in the above step.

Reviewer: ```suggestion
* Now run the .jar file created in the above step.
```

```bash
$ java -jar target/bin/employee_details_service.jar
```
You will be prompted to enter the secret you used for encrypting the FTP username and password.

```bash
ballerina: enter secret for config value decryption:
```
Now we can see that the service has started on port 9090. 

* Let’s access this service by executing the following curl command.

```bash
$ curl http://localhost:9090/organization/employees
```

> We shall get a json response like the following.

Reviewer: ```suggestion
We receive a JSON response similar to the following.
```
